qadence_protocols/measurements/shadow.py

```diff
@@ -4,90 +4,159 @@
+        return self.options
+
+    def number_of_samples_and_groups(
+        self, observables: list[AbstractBlock] = list()
+    ) -> tuple[int, ...]:
+        """Return the shadow size and number of groups for a list of observables.
+
+        Args:
+            observables (list[AbstractBlock], optional): List of observables.
+                Defaults to list().
+
+        Returns:
+            tuple[int, ...]: the shadow size and number of groups
+        """
+        return number_of_samples(
+            observables=observables,
+            accuracy=self.options["accuracy"],
+            confidence=self.options["confidence"],
+        )
+
+    def reconstruct_state(self, snapshots: Tensor) -> Tensor:
+        """Reconstruct the state from the snapshots.
+
+        Args:
+            snapshots (Tensor): Snapshots of size
+                (batch_size, shadow_size, 2**n, 2**n).
+
+        Returns:
+            Tensor: Reconstructed state.
+        """
+        N = snapshots.shape[1]
+        return snapshots.sum(axis=1) / N
+
+    def get_snapshots(
+        self,
+        model: QuantumModel,
+        param_values: dict[str, Tensor] = dict(),
+        state: Tensor | None = None,
+    ) -> Tensor:
+        """Obtain snapshots from the measurement data.
+
+        Args:
+            model (QuantumModel): Quantum model instance.
+            param_values (dict[str, Tensor], optional): Parameter values. Defaults to dict().
+            state (Tensor | None, optional): Input state. Defaults to None.
+
+        Returns:
+            Tensor: Snapshots for a input circuit model and state.
+                The shape is (batch_size, shadow_size, 2**n, 2**n).
+        """
+        if self.measurement_data is None:
+            self.measure(model, list(), param_values, state)
+
+        unitaries_ids, bitstrings = self.measurement_data  # type: ignore[misc]
+        return compute_snapshots(bitstrings, unitaries_ids, local_shadow)
+
+    def measure(
+        self,
+        model: QuantumModel,
+        observables: list[AbstractBlock] = list(),
+        param_values: dict[str, Tensor] = dict(),
+        state: Tensor | None = None,
+    ) -> MeasurementData:
+        """Obtain measurement data from a quantum program for classical shadows.
+
+        Args:
+            model (QuantumModel): Quantum model instance.
+            observables (list[AbstractBlock], optional): List of observables. Defaults to list().
+            param_values (dict[str, Tensor], optional): Parameter values. Defaults to dict().
+            state (Tensor | None, optional): Input state. Defaults to None.
+
+        Returns:
+            MeasurementData: Measurement data as locally sampled pauli unitaries and
+                samples from the circuit
+                rotated according to the locally sampled pauli unitaries.
+        """
+
+        circuit = model._circuit.original
+        shadow_size = self.options["shadow_size"]
```

Comment:
Have you verified these options are present here ?

Reply:
What I meant is I didn't see where the `verify_options` is triggered before coming here. Maybe I just missed it.